packages/react-query/src/useBaseQuery.ts

```diff
@@ -65,37 +66,75 @@ export function useBaseQuery<
 
   useClearResetErrorBoundary(errorResetBoundary)
 
-  const [observer] = React.useState(
-    () =>
-      new Observer<TQueryFnData, TError, TData, TQueryData, TQueryKey>(
-        queryClient,
-        defaultedOptions,
-      ),
-  )
+  /**
+   * It can take much time between useLazyRef and useSyncExternalStore callback.
+   * Let's assume, we have two components (A and B), with the same query inside.
+   * There is a potential situation, when the component B can subscribe to changes
+   * after these changes occured in the component A.
+   *
+   * You can get more info here: https://github.com/TanStack/query/issues/5443
+   */
+  const initiallyUnsubscribe = React.useRef<() => void>()
+  const needsToFlushStoreChange = React.useRef(false)
+  const observerRef = useLazyRef(() => {
+    const observer = new Observer<
+      TQueryFnData,
+      TError,
+      TData,
+      TQueryData,
+      TQueryKey
+    >(queryClient, defaultedOptions)
+
+    initiallyUnsubscribe.current = observer.subscribe(() => {
+      needsToFlushStoreChange.current = true
+    })
```

Comment:
I thought the agreement was to use layout effect instead ? I'm pretty sure it's against the rules of react to create a subscription inside an initializer, which should be pure.

Reply:
It was an attempt to create a subscription on the first render) But yeah, you're right, it's quite tricky